File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.go
func maxLength(arr []string) int {
    

    c := []uint32{}
    ans := 0 
    for _ , s := range arr { 
        var mask uint32
        for _ , c := range s { 
            mask = mask | 1 << (c-'a')
        }
        if len(s) != bits.OnesCount32(mask) {
            continue }
        c = append(c, mask)
    }
    dfs(c, 0, 0, &ans)
    return ans 
    }


func dfs(c []uint32 , ix int , mask uint32 , ans *int){

        *ans = Max(*ans , bits.OnesCount32(mask))
        for i := ix ; i < len(c); i++ { 
            if mask & c[i] == 0 { 
                dfs(c , i+1 , mask | c[i] , ans)
            }
        }
        return 
    }

func Max(a,b int) int{ 
        if a > b { 
            return a } 
    return b 
    }

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    int maxLength(vector<string>& arr) {
        
    vector<bitset<26>> dp = {bitset<26>()};
    int ans = 0;
    for(auto& s: arr){
        bitset<26> bit ; 
        for(char c : s){
            bit.set(c-'a'); }
        int n = bit.count(); 
        if( n < s.size()) continue ; 

        for( int i= dp.size() -1 ; i>= 0 ; --i){
            bitset bit2 = dp[i] ;
            if ((bit2 & bit).any()) continue ; 
            dp.push_back(bit2|bit) ;
            ans = max(ans, (int)bit2.count()+n);
        }
    }
    return ans; 
    }
};